# Code and data repository for Selechnik et al 2020

## Inherent population structure determines the importance of filtering parameters for reduced representation sequencing analyses

D. Selechnik <sup>1,2*</sup>, M.F. Richardson <sup>3,4</sup>, M.K. Hess <sup>5</sup>, A.S. Hess <sup>5</sup>, K.G. Dodds <sup>5</sup>, M. Martin <sup>4</sup>, T.C. Chan <sup>2</sup>, A.P.A. Cardilini <sup>4</sup>, C. Sherman <sup>4</sup>, R. Shine <sup>1</sup>, L.A. Rollins <sup>2</sup>

<sup>1</sup> School of Life and Environmental Sciences (SOLES), University of Sydney 2006;
<sup>2</sup> Evolution & Ecology Research Centre, School of Biological, Earth and Environmental Sciences, University of New South Wales, Sydney, NSW 2052;
<sup>3</sup> Deakin Genomics Centre, School of Life and Environmental Sciences, Deakin University, Locked Bag 20000, Geelong, VIC, Australia 3216;
<sup>4</sup> Centre for Integrative Ecology, School of Life and Environmental Sciences, Deakin University, Locked Bag 20000, Geelong, VIC, Australia 3216;
<sup>5</sup> AgResearch Limited, Invermay Agricultural Centre, Puddle Alley, Mosgiel, New Zealand.


> Compiled code to run our analyses can be found under the **code** subdirectory in the shell script *Code_Supplemental.sh*
>> its probably best to run the annotated code chunks separately instead of just executing the whole script


> The simulated data that serves as a basis for our filtering comparisons can be found in the **data** subdirectory
>> We provide the simulated data for all the iterations at different depth thresholds, as well as the corresponding true genotypes.


If you use any of the code or data please cite our publication:

ADD CITATION
